pkg/authorization/deep_sar.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorization
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kuser "k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/rest"
+)
+
+type deepSARKeyType int
+
+const (
+	deepSARKey deepSARKeyType = iota
+)
+
+var (
+	deepSARHeader = "X-Kcp-Internal-Deep-SubjectAccessReview"
+)
+
+// WithDeepSARConfig returns a clone of the input rest.Config
+// with an additional header making SARs to be deep.
+func WithDeepSARConfig(config *rest.Config) *rest.Config {
```

Comment:
nit: we hit this in another PR - let's prefer not to do the copy here and let the caller decide whether or not mutating the original object is OK

Reply:
i can remove it, how should it look like, can you point me to the PR?

Reply:
Just don't do the `rest.CopyConfig()` in this helper, assume that if the user wants to copy, they will have done so before passing you the config to mutate.

Reply:
fixed